Llava Onevision: add model

zucchini-nlp opened the pull request:
# What does this PR do?

Adds llava onevision models (https://arxiv.org/abs/2408.03326) which are basically same as LLaVa-NeXT with have a different vision tower and image grid pinpoints

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32673). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts commented:
Thanks for adding this model @zucchini-nlp! 

If the model has its own paper - we should add it as its own model and leverage the diff converter for the model addition if changes are needed

NielsRogge commented:
Agree with @amyeroberts, which is more in line with the [philosophy](https://huggingface.co/blog/transformers-design-philosophy)

codeb2cc commented:
@zucchini-nlp You've hardcoded a path in modeling_llava_next.py
```
        if input_ids.shape[1] > 1:
            torch.save(inputs_embeds, "/raid/raushan/inputs_embeds_our.pt")
```

zucchini-nlp commented:
Yes, will remove that. I am making a completely new modeling folder for Onevision currently

zucchini-nlp committed:
working version

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
update

zucchini-nlp committed:
tests

zucchini-nlp committed:
update docs

zucchini-nlp committed:
codestyle

zucchini-nlp committed:
Merge branch 'huggingface:main' into llava-onevision

zucchini-nlp committed:
add more tests

zucchini-nlp committed:
add returns for docs

zucchini-nlp commented:
@amyeroberts Ready for review. Added a new model LLaVa-Onevision. I didn't use diff because I realized that copying everything from llava-next is not a good decision. The current code incorporates all the recent refactors we've been working on and supports `GenerationTesterMixin` 🎉  Apart from that it can support static cache, though not compile due to dynamic control.

I added two processing files, one for image and another for video, since I've been planning to go over video-image LLMs and separate processing into two files. And later add more utility functions to the `BaseVideoProcessor` class. But in general, there's no difference between them except for that videos can be list of 4D tensors and we need one more loop over frame dimension.

zucchini-nlp committed:
clean up

amyeroberts commented on src/transformers/models/llava_onevision/modeling_llava_onevision.py:
```diff
@@ -0,0 +1,675 @@
+                module.weight.data[module.padding_idx].zero_()
+
+
+LLAVA_ONEVISION_INPUTS_DOCSTRING = r"""
```
This isn't used

amyeroberts commented on src/transformers/models/llava_onevision/modeling_llava_onevision.py:
```diff
@@ -0,0 +1,675 @@
+"""
+
+
+class LlavaOnevisionForConditionalGeneration(LlavaOnevisionPreTrainedModel):
```
missing decorator to add class docstring

amyeroberts commented on src/transformers/models/llava_onevision/modeling_llava_onevision.py:
```diff
@@ -0,0 +1,675 @@
+        feature_lens = torch.tensor(feature_lens, dtype=torch.long, device=image_features.device)
+        return image_features, feature_lens
+
+    def forward(
```
missing dosctring decorator

amyeroberts commented on src/transformers/models/llava_onevision/modeling_llava_onevision.py:
```diff
@@ -0,0 +1,675 @@
+
+LLAVA_ONEVISION_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
```
Order here doesn't match signature

amyeroberts commented on src/transformers/models/llava_onevision/modeling_llava_onevision.py:
```diff
@@ -0,0 +1,675 @@
+                # [batch_size*frames*num_patches, num_channels, height, width]
+                pixels = torch.cat(_pixel_values_list, dim=0)
+            elif pixels.dim() != 4:
+                raise ValueError(f"pixel_values_videos of shape {pixels.shape}, expect to be of 4 or 5 dimensions")
```
this is error a bit misleading for the user: 
* `pixels` here could he for `pixel_values_videos` or `pixel_values`
* I don't think `pixel_values_videos` should be of dim 4 or 5

zucchini-nlp replied:
Yes, maybe we'll need to raise an error before flattening to check number of dims is correct

amyeroberts commented on src/transformers/models/llava_onevision/processing_llava_onevision.py:
```diff
@@ -0,0 +1,275 @@
+    [`~LlavaOnevisionVideoProcessor.__call__`], [`~LlavaNextProcessor.__call__`] and [`~LlavaNextProcessor.decode`] for more information.
+
+    Args:
+        video_processor ([`LlavaOnevisionVideoProcessor`], *optional*):
+            The video processor is a required input.
+        image_processor ([`LlavaNextImageProcessor`], *optional*):
+            The image processor is a required input.
+        tokenizer ([`LlamaTokenizerFast`], *optional*):
+            The tokenizer is a required input.
```
I think it would be better to preserve the default order of processing classes here - so tokenizer, image processor and then video processor. cc @molbap for this opinion

molbap replied:
yes, agreed! even though the model is suited for vision, we want to preserve that order across the lib, be it for processors or their kwargs

zucchini-nlp replied:
oke, actually I added another model with this order so I'll open a PR to fix that one also

amyeroberts commented on src/transformers/models/llava_onevision/processing_llava_onevision.py:
```diff
@@ -0,0 +1,275 @@
+        of the above two methods for more information.
+
+        Args:
+            text (`str`, `List[str]`, `List[List[str]]`):
```
Doesn't match signature order

amyeroberts commented on src/transformers/models/llava_onevision/processing_llava_onevision.py:
```diff
@@ -0,0 +1,275 @@
+        if images is not None:
+            image_inputs = self.image_processor(images, **output_kwargs["images_kwargs"])
+
+            image_sizes = to_py_obj(image_inputs["image_sizes"])
```
Rather than create a new object - we could just create an iterator and just grab the next value in the while loop

zucchini-nlp replied:
oke, can do so

amyeroberts commented on src/transformers/models/llava_onevision/processing_llava_onevision.py:
```diff
@@ -0,0 +1,275 @@
+            image_inputs = self.image_processor(images, **output_kwargs["images_kwargs"])
+
+            image_sizes = to_py_obj(image_inputs["image_sizes"])
+            height, width = get_image_size(to_numpy_array(image_inputs["pixel_values"][0][0]))
```
We might need to pass in the data format here

zucchini-nlp replied:
Hmm, is that for cases when height or width can have same size as `num_channels` ? I think that's very unlikely, even for tests

But passing format shouldn't be hard, from `images_kwargs`. That will work only if users specifically asked for certain data format, otherwise we'll have to infer

amyeroberts commented on src/transformers/models/llava_onevision/processing_llava_onevision.py:
```diff
@@ -0,0 +1,275 @@
+            num_image_tokens = self.num_image_tokens
+            prompt_strings = []
+
+            for sample in text:
+                while self.video_token in sample:
+                    image_size_list = image_sizes.pop(0)
+                    orig_height, orig_width = image_size_list[0]
+                    num_image_tokens = self._get_number_of_features(orig_height, orig_width, height, width)
+                    if self.vision_feature_select_strategy == "default":
+                        num_image_tokens -= 1
+
+                    sample = sample.replace(self.video_token, "<placeholder>" * num_image_tokens * num_frames, 1)
+                prompt_strings.append(sample)
```
Common logic can be abstracted out between images and videos

amyeroberts commented on src/transformers/models/llava_onevision/processing_llava_onevision.py:
```diff
@@ -0,0 +1,275 @@
+        )
+        scale_height, scale_width = height_best_resolution // height, width_best_resolution // width
+
+        patches_height = int(math.sqrt(self.num_image_tokens))
+        patches_width = int(math.sqrt(self.num_image_tokens))
```
nit - we don't need to worry about shared references for an int here 

```suggestion
        patches_height = patches_width = int(math.sqrt(self.num_image_tokens))
```

amyeroberts commented on tests/models/llava_onevision/test_modeling_llava_onevision.py:
```diff
@@ -0,0 +1,523 @@
+                out_embeds = model(inputs_embeds=inputs_embeds, **inputs)[0]
+            self.assertTrue(torch.allclose(out_embeds, out_ids))
+
+    @unittest.skip(
+        reason="This architecure seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing(self):
+        pass
+
+    @unittest.skip(
+        reason="This architecure seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing_use_reentrant(self):
+        pass
+
+    @unittest.skip(
+        reason="This architecure seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing_use_reentrant_false(self):
+        pass
+
+    @unittest.skip(reason="Feedforward chunking is not yet supported")
+    def test_feed_forward_chunking(self):
+        pass
+
+    @unittest.skip(reason="CPU offload is not yet supported")
+    def test_cpu_offload(self):
+        pass
```
Have these been run without the skips?

zucchini-nlp replied:
no, copied from llava-next which is mostly same architecture as the current model. Let me see if these were resolved somehow or still have to be skipped

amyeroberts commented on docs/source/en/model_doc/llava_onevision.md:
```diff
@@ -0,0 +1,312 @@
+
+### Video inference
+
+LLaVa-Onevision can perform inference with multiple images as input, where images either belong to the same prompt or different prompts (in batched inference). Here is how you can do it:
```
Explanation here talks about images - but this section is about video.

amyeroberts commented on docs/source/en/model_doc/llava_onevision.md:
```diff
@@ -0,0 +1,312 @@
+
+<Tip warning={true}>
+
+- Llava-Next uses different number of patches for images and thus has to pad the inputs inside modeling code, aside from the padding done when processing the inputs. The default setting is "left-padding" if model is in `eval()` mode, otherwise "right-padding".
```
To be updated - this isn't llava-next

amyeroberts commented on tests/generation/test_utils.py:
```diff
@@ -459,7 +459,8 @@ def test_greedy_generate_dict_outputs(self):
                 self.assertIsInstance(output_generate, GreedySearchDecoderOnlyOutput)
 
-            self._check_outputs(output_generate, input_ids, model.config)
+            config = model.config if not hasattr(model.config, "text_config") else model.config.text_config
```
This feels brittle and makes a strong assumption about the structure of our configs it would be good to have @gante's view on this here. 

I'd suggest moving this handling into `_check_outputs` so the logic change only has to be in on place rather than every `_check_outputs` call, which already has conditional logic on e.g. `config.is_encoder_decoder` within the utility

gante replied:
We are starting to have more places where we need to pull the decoder part of a cache in a composite model:
- in the generation config
- in while saving a pretrained config, to ensure it doesn't contain generate flags [open PR]
- here

Note that they are all `generate`-related. In this case, the shape of the `generate` outputs depend on the decoder, so we need to pull the decoder config.

@amyeroberts I was thinking of abstracting this logic into `GenerationMixin`, in a function that pulls the decoder config given a config. WDYT? :)

amyeroberts replied:
I'd be pro! This should help encapsulate the logic without having to handle it everywhere else.

gante replied:
Cool, I'll open a PR after https://github.com/huggingface/transformers/pull/32659 gets merged (to avoid conflicts :) )

zucchini-nlp replied:
Great, will port these to `_check_output` until the PR is merged

gante replied:
FYI opening a PR to add the generic function AND fixing `_check_output ` (#33212 also needs it :D )

gante replied:
https://github.com/huggingface/transformers/pull/33219

amyeroberts reviewed (commented):
Thanks for adding! 

Main comment is about the new logic to use `config.text_config`

zucchini-nlp commented on src/transformers/cache_utils.py:
```diff
@@ -1009,6 +1009,10 @@ def __init__(self, config: PretrainedConfig, max_batch_size: int, max_cache_len:
         super().__init__()
         self.max_batch_size = max_batch_size
         self.max_cache_len = config.max_position_embeddings if max_cache_len is None else max_cache_len
+
+        if hasattr(config, "text_config"):  # in case of composite models
+            config = config.text_config
+
```
We can do it inside cache class also, and in that case handling of composite models will move to cache initialization. I see comment from @gante about a helper function to get decoder/composite models' configs. In that case we can leave it here imo, and after @gante merged a PR modify this and other places to `self.get_text_config(config)`

zucchini-nlp commented on src/transformers/generation/utils.py:
```diff
@@ -2365,7 +2365,11 @@ def _dola_decoding(
 
         # prepare layers for DoLa decoding
-        final_layer = self.config.num_hidden_layers
+        final_layer = (
+            self.config.text_config.num_hidden_layers
+            if hasattr(self.config, "text_config")
+            else self.config.num_hidden_layers
+        )
```
Hmm, let me see if we can rely on smth else and not config, and check out the case if a `encoder-decoder` config is used. Prob we'll have to handle it same way as above, via a helper function

zucchini-nlp commented on src/transformers/models/llava_onevision/modeling_llava_onevision.py:
```diff
@@ -0,0 +1,675 @@
+
+            # if videos are being processed, we can flatten frame dim and treat it as a huge batch of images
+            if pixels.dim() == 6:
+                sizes = [imsize for size_list in sizes for imsize in size_list]
```
Right, will raise an error then, because `sizes` are required to unpad if corresponding pixels are passed

zucchini-nlp committed:
Update src/transformers/models/llava_onevision/processing_llava_onevision.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
updates

zucchini-nlp committed:
codestyle

zucchini-nlp committed:
Merge branch 'main' into llava-onevision

zucchini-nlp committed:
style

zucchini-nlp commented:
@amyeroberts should be ready! I left the `config vs text_config` when applicable until @gante merged his PR, after that those can be replaced by a call to something like `self.get_config()`

zucchini-nlp committed:
shouldn't be reversed

zucchini-nlp committed:
[run-slow] llava_onevision

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into llava-onevision

zucchini-nlp committed:
[run-slow] llava_onevision

zucchini-nlp commented:
Failing tests are OOM or not related

zucchini-nlp committed:
Merge branch 'huggingface:main' into llava-onevision

amyeroberts commented:
> Failing tests are OOM or not related

If the tests are failing with OOM, this is still an issue which means we can't merge them into main, unfortunately

zucchini-nlp commented:
Oke, I'll load them in lower precision. Wondering if the GPU memory is cleared after each slow test as the model is only 0.8B parameters and the GPU apparently has 14GB memory

amyeroberts commented:
@zucchini-nlp hmmm - that's funny. @ydshieh will know more than me about the mem management here!

Luodian commented:
May I know if the video frames are correctly processed (they need to be bilinear interpolated to 196 tokens each frame). I quickly checked the code and didnt see such operations, may wondering if that will cause the OOM problem?

zucchini-nlp commented:
Hey @Luodian , thanks for reviewing the PR! Yes, tye implementation doesn't do interpolation. I saw you guys mentioned it in the paper but the demo notebook didn't lead to the path with interpolation, so I thought it's from llava-next-video. 

Prob it was a typo in demo notebook, passing "modality" arg triggers interpolation. I'll add it then, thanks ;)

Luodian commented:
> Hey @Luodian , thanks for reviewing the PR! Yes, tye implementation doesn't do interpolation. I saw you guys mentioned it in the paper but the demo notebook didn't lead to the path with interpolation, so I thought it's from llava-next-video.
> 
> Prob it was a typo in demo notebook, passing "modality" arg triggers interpolation. I'll add it then, thanks ;)

Yes, there's mistake on `modality` in our previous notebook demo. Really nice and also thanks to see you made such great contribution to the community.

zucchini-nlp committed:
add pooling in videos

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into llava-onevision

zucchini-nlp committed:
[run-slow] llava_onevision

zucchini-nlp committed:
num-logits-to-keep

zucchini-nlp committed:
[run-slow] llava_onevision

zucchini-nlp committed:
[run-slow] llava_onevision

zucchini-nlp commented:
@amyeroberts done! The tests are passing now, including slow + added the extra pooling layer for videos

amyeroberts commented on tests/test_modeling_common.py:
```diff
@@ -4831,8 +4831,11 @@ def test_forward_with_num_logits_to_keep(self):
-            vocab_size = config.vocab_size
+            vocab_size = config.text_config.vocab_size if hasattr(config, "text_config") else config.vocab_size
             model = model_class(config).to(device=torch_device).eval()
+            _ = inputs.pop(
+                "labels", None
+            )  # some models have labels but `num_logits_to_keep` should not be used in train mode
```
nit - comment above relevant line avoids unnecessary line splitting 

```suggestion
            # some models have labels but `num_logits_to_keep` should not be used in train mode
            _ = inputs.pop("labels", None)
```

amyeroberts commented on src/transformers/models/llava_onevision/processing_llava_onevision.py:
```diff
@@ -0,0 +1,276 @@
+            num_video_tokens = num_frames * pooled_height_width * pooled_height_width
+            text = [sample.replace(self.video_token, self.video_token * num_video_tokens) for sample in text]
+
+        # Padding side can be in TextKwargs but is not accepted by the tokenizer
+        _ = output_kwargs["text_kwargs"].pop("padding_side", None)
```
We should find a way to handle this better as we don't want to have to add this for all the processors cc @molbap

zucchini-nlp replied:
I think we can just remove `padding_side` from TextKwargs, which I believe is already done in one of the PRs for standardizaing

amyeroberts reviewed (approved):
Looks great - thanks for all the work adding!

zucchini-nlp committed:
Update tests/test_modeling_common.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp commented:
Will need some changes in video processing as discussed here (https://github.com/LLaVA-VL/LLaVA-NeXT/issues/144). Comes out the demo notebook was buggy and videos should be processed in a different way. Will work on it when I get reply from authors on what should be the correct way

zucchini-nlp committed:
video matched orig impl

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into llava-onevision

amyeroberts commented:
@zucchini-nlp Thanks for the update! I'll mark myself a re-requesting for review then as this will be a significat change

zucchini-nlp committed:
fix tests

zucchini-nlp committed:
chat template was modified

zucchini-nlp commented:
@amyeroberts yes, the tests should be green now. Merging `main` somehow rewinded my local changes

amyeroberts commented on docs/source/en/model_doc/llava_onevision.md:
```diff
@@ -0,0 +1,312 @@
+videos.*
+
+Tips:
+
```
nit 
```suggestion
```

amyeroberts commented on docs/source/en/model_doc/llava_onevision.md:
```diff
@@ -0,0 +1,312 @@
+
+## Overview
+
+The LLaVA-Onevision model was proposed in [LLaVA-OneVision: Easy Visual Task Transfer](https://arxiv.org/abs/2408.03326) by <Bo Li, Yuanhan Zhang, Dong Guo, Renrui Zhang, Feng Li, Hao Zhang, Kaichen Zhang, Yanwei Li, Ziwei Liu, Chunyuan Li
```
Could you add a short description of the model here? This is what will be used in the release notes. A good example is the [music gen page](https://huggingface.co/docs/transformers/index)

zucchini-nlp replied:
done

amyeroberts reviewed (approved):
Looks great - thanks for all the work iterating on this! 

Only thing is to add a short description of the model and resolve the failing SDPA test before merge

zucchini-nlp committed:
Update docs/source/en/model_doc/llava_onevision.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
add morer info in the doc page

zucchini-nlp committed:
Merge branch 'main' into llava-onevision

zucchini-nlp commented:
oke, will merge it now as the model works. The only thing I am struggling now is ONNX conversion which is not related to the PR

mylesgoose commented:
@zucchini-nlp 

hello any idea how to get a llama model working with this pr. mylesgoose/Llama-3.1-Minitron-4B-Llava-Nvidia-siglip-ov regards myles I have tried adding the similar files to the huggingface repo as per the qwen model which works. there is small issue I think because  the 4b model is a stripped down 8b model with weight sizes.

mylesgoose commented:
@zucchini-nlp  no I made the model using your git repo repo llava one vision. it's a llama 4b model. and this is a feature request. I trained it.  because the model is only supported of qwen type models. at present this transformers can't understand  my weight are smaller. I am not sure exactly  how to modify this one vision  transformers to load the llama model type. can you have a look at the huggingface repo.and you will see what I mean.

zucchini-nlp commented:
@mylesgoose I am not sure I understand you. The repo shows a trained model with a different LM backbone and has inference script with llava-vl repo code. So I guess you did training and can do generation with llava-vl repo

In transformers we already support any LM backbone, and all you need is to convert the weights, which has several conversion scripts for different llava models. So this is not a feature request. In case you still believe we can help you by adding/modifying llava code, please open an issue and provide minimal script to see what you want to achieve and what output is expected from running it 🤗

mylesgoose commented:
Okay thanks @zucchini-nlp  I understand now. I found that script and ran one to generate the bin file and that worked. however the script does not work on the llama model. so I will modify it tomorrow and get it to work and then return working adjusted one. I think it also does not add tokens for video etc also. or check if pad token exists prior to creating one. and created meta tensors on devices and could not use in latter step. will figure it out. thanks allot.

mylesgoose commented:
This seems to have worked. @zucchini-nlp

```python
import argparse
import torch
from transformers import (
    AddedToken,
    AutoConfig,
    AutoImageProcessor,
    AutoTokenizer,
    LlavaConfig,
    LlavaForConditionalGeneration,
    LlavaProcessor,
    SiglipVisionConfig,
)

EPILOG_TXT = """Example:
    python convert_llava_weights_to_hf.py \\
      --text_model_id mylesgoose/Llava-Llama-3.1-Minitron-4B-Nvidia-siglip-ov \\
      --vision_model_id google/siglip-so400m-patch14-384 \\
      --output_hub_path mylesgoose/Llama-3.1-Minitron-4B-Llava-Nvidia-siglip-ov-hf \\
      --old_state_dict_id /home/myles/LLaVA-NeXT/tmp/hf_models/llava-Llama-3.1-Minitron-4B/model_state_dict.bin

Example for creating the old state dict file with Python:

    import torch
    from llava.model.language_model.llava_llama import LlavaLlamaForCausalLM

    # Load model
    kwargs = {"device_map": "auto", "torch_dtype": torch.float16}
    model = LlavaLlamaForCausalLM.from_pretrained("mylesgoose/Llava-Llama-3.1-Minitron-4B-Nvidia-siglip-ov", low_cpu_mem_usage=True, **kwargs)

    # Load vision tower
    model.get_vision_tower().load_model()

    # Save state dict
    torch.save(model.state_dict(), "/home/myles/LLaVA-NeXT/tmp/hf_models/llava-Llama-3.1-Minitron-4B/model_state_dict.bin")
"""

KEYS_TO_MODIFY_MAPPING = {
    "model.vision_tower.vision_tower.": "vision_tower.",
    "model.vision_tower.": "vision_tower.",
    ".vision_resampler": "",  # Not used in the original implementation
    "model.mm_projector": "multi_modal_projector",
    "model": "model.model",
    "vision_model.model": "vision_model",
    "lm_head": "language_model.lm_head",
    "model.model": "language_model.model",
    "multi_modal_projector.0": "multi_modal_projector.linear_1",
    "multi_modal_projector.2": "multi_modal_projector.linear_2",
}

def convert_state_dict_to_hf(state_dict):
    new_state_dict = {}
    for key, value in state_dict.items():
        # Skip keys ending with .inv_freq (if present)
        if key.endswith(".inv_freq"):
            continue
        # Modify keys based on the mapping
        for key_to_modify, new_key in KEYS_TO_MODIFY_MAPPING.items():
            if key_to_modify in key:
                key = key.replace(key_to_modify, new_key)
        new_state_dict[key] = value
    return new_state_dict

def convert_llava_llama_to_hf(text_model_id, vision_model_id, output_hub_path, old_state_dict_id):
    torch.set_default_dtype(torch.bfloat16)
    text_config = AutoConfig.from_pretrained(text_model_id)

    tokenizer = AutoTokenizer.from_pretrained(text_model_id)
    tokenizer.add_tokens(AddedToken("<image>", special=True, normalized=False), special_tokens=True)
    tokenizer.add_tokens(AddedToken("<video>", special=True, normalized=False), special_tokens=True)
    if tokenizer.pad_token_id is None:
        tokenizer.add_special_tokens({"pad_token": "<|finetune_right_pad_id|>"})

    image_processor = AutoImageProcessor.from_pretrained(vision_model_id)
    processor = LlavaProcessor(tokenizer=tokenizer, image_processor=image_processor)

    # Set vision_config based on the vision model
    vision_config = SiglipVisionConfig.from_pretrained(vision_model_id)

    # Adjust the vision_config to match your model's parameters
    vision_config.hidden_size = 1152
    vision_config.num_attention_heads = 16
    vision_config.num_hidden_layers = 24
    vision_config.intermediate_size = 4304
    vision_config.patch_size = 14
    vision_config.image_size = 384
    vision_config.vision_use_head = False

    config = LlavaConfig(
        text_config=text_config,
        vision_config=vision_config.to_dict(),
    )

    # Set other configuration parameters as needed
    config.pad_token_id = tokenizer.pad_token_id
    config.image_token_index = tokenizer.convert_tokens_to_ids("<image>")
    config.video_token_index = tokenizer.convert_tokens_to_ids("<video>")
    config.vision_feature_layer = -2  # Adjust if necessary
    config.vision_tower_pretrained = vision_model_id

    # Instantiate the model
    model = LlavaForConditionalGeneration(config)

    # Load the state dictionary from the provided path
    state_dict_path = old_state_dict_id
    state_dict = torch.load(state_dict_path, map_location="cpu")

    # Convert state dict to match Hugging Face model's expected keys
    state_dict = convert_state_dict_to_hf(state_dict)

    # Load the state dictionary into the model
    missing_keys, unexpected_keys = model.load_state_dict(state_dict, strict=False)
    print(f"Missing keys: {missing_keys}")
    print(f"Unexpected keys: {unexpected_keys}")

    # Resize token embeddings to accommodate new tokens
    model.resize_token_embeddings(len(tokenizer))

    # Get the updated sizes
    pre_expansion_embeddings = model.language_model.model.embed_tokens.weight.data
    vocab_size = pre_expansion_embeddings.size(0)
    new_vocab_size = model.language_model.model.embed_tokens.weight.size(0)
    num_new_tokens = new_vocab_size - vocab_size

    print(f"Original vocab size: {vocab_size}")
    print(f"New vocab size: {new_vocab_size}")
    print(f"Number of new tokens: {num_new_tokens}")

    # Initialize new token embeddings only if there are new tokens
    if num_new_tokens > 0:
        mu = torch.mean(pre_expansion_embeddings, dim=0).float()
        n = pre_expansion_embeddings.size(0)
        sigma = ((pre_expansion_embeddings - mu).T @ (pre_expansion_embeddings - mu)) / n
        dist = torch.distributions.multivariate_normal.MultivariateNormal(
            mu.cpu(), covariance_matrix=1e-5 * sigma.cpu()
        )

        new_embeddings = torch.stack(
            [dist.sample() for _ in range(num_new_tokens)], dim=0
        ).to(pre_expansion_embeddings.device)
        model.language_model.model.embed_tokens.weight.data[vocab_size:] = new_embeddings
        model.language_model.lm_head.weight.data[vocab_size:] = new_embeddings
    else:
        print("No new tokens to add to embeddings.")

    # Save the converted model and processor
    model.push_to_hub(output_hub_path)
    processor.push_to_hub(output_hub_path)

def main():
    parser = argparse.ArgumentParser(
        epilog=EPILOG_TXT,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument(
        "--text_model_id",
        default="mylesgoose/Llava-Llama-3.1-Minitron-4B-Nvidia-siglip-ov",
        help="Hub location of the text model",
    )
    parser.add_argument(
        "--vision_model_id",
        default="google/siglip-so400m-patch14-384",
        help="Hub location of the vision model",
    )
    parser.add_argument(
        "--output_hub_path",
        default="mylesgoose/Llama-3.1-Minitron-4B-Llava-Nvidia-siglip-ov-hf",
        help="Location on the hub of the converted model",
    )
    parser.add_argument(
        "--old_state_dict_id",
        default="/home/myles/LLaVA-NeXT/tmp/hf_models/llava-Llama-3.1-Minitron-4B/model_state_dict.bin",
        help="Path to the raw state dict of the original model. The filename needs to be `model_state_dict.bin`",
    )
    args = parser.parse_args()
    convert_llava_llama_to_hf(
        args.text_model_id,
        args.vision_model_id,
        args.output_hub_path,
        args.old_state_dict_id
    )

if __name__ == "__main__":
    main()
```
